### ISSUE TITLE
Dokku to use only one IP and co-exist peacefully with existing Apache server

### ISSUE THREAD

**Issue by babakness** [dokku/dokku#3192]
Hello,

There is a server with two IPs going into the host. Apache is configured to use one of these IPs only. This is done simply by configuring the `listen` directive. I'd like have `dokku` manage containers on the other IP without interfering with Apache. However, it seems by default, the `docker` containers are listening to any IP going into that host. 

I've tried editing `/etc/docker/daemon.json` per

https://docs.docker.com/v17.09/engine/userguide/networking/default_network/binding/

> Or if you always want Docker port forwards to bind to one specific IP address, you can edit your system-wide Docker server settings and add the option --ip=IP_ADDRESS. 

I do this and issue `service docker restart`. 

I've also tried
1. Rebooting the host after the change
2. Giving a non-existing IP as a fail test

But to no avail. The containers all listen to any ip going into that host on any interface.

This is occurring on a droplet on digital ocean. The host is an Ubuntu server, 18.04 LTS. Dokku version 0.12.5

**Comment by josegonzalez** [comment]
Can you include *all* of the information we ask for in our [issue template](https://github.com/dokku/dokku/blob/master/ISSUE_TEMPLATE.md)? It is helpful for debugging problems. Thanks.

As well, I would like to see the trace output of a deploy that doesn't work the way you want it to.

**Comment by babakness** [comment]
Hi, 

I should mention this isn't about a specific container but for all containers. 

> dokku report APP_NAME

This isn't an issue about a specific app, but about confining all apps to listen to only one IP coming into the host.

> docker inspect CONTAINER_ID (if applicable): (BEWARE: docker inspect will print environment variables for some commands, be sure you're not exposing any sensitive information when posting issues. You may replace these values with XXXXXXX):

ditto -- N/A

> cat /home/dokku/<app>/nginx.conf (if applicable):

ditto -- N/A

> Link to the exact repository being deployed (if possible/applicable):

ditto, sample app used to test

```javascript
const express = require('express')
const app = express()
app.get('/', (req, res) => res.send('Hello World!'))
app.listen(process.env.PORT || 3000, () => console.log('Example app listening on port' + process.env.PORT))
```

> If a deploy is failing or behaving unexpectedly:

Not failing, just not being confined to a specific IP. Listening to all IPs on machine rather than specific IP on specific interface

> Application name

node_sample
nginx_sample (using .static buildpack)

> The type of application being deployed (node, php, python, ruby, etc.)

node, ruby (from tutorial), static (static buildpack)

> If using buildpacks, which custom buildpacks are in use

I don't believe this is relevant but, .static

> If using a Dockerfile, the contents of that file

N/A

> If it exists, the contents of your Procfile.

N/A

> Output of failing Dokku commands after running dokku trace on (BEWARE: trace on will print environment variables for some commands, be sure you're not exposing any sensitive information when posting issues. You may replace these values with XXXXXXX):

Nothing is failing


---

Allow me to re-iterate, this is an issue about binding incoming connections for ALL containers to a specific IP.

I've followed this 

https://docs.docker.com/v17.09/engine/userguide/networking/default_network/binding/

> Or if you always want Docker port forwards to bind to one specific IP address, you can edit your system-wide Docker server settings and add the option --ip=IP_ADDRESS. 

However, it doesn't work here. It should work, see this stackoverflow 

https://stackoverflow.com/questions/50421245/docker-not-binding-to-specified-ip

Thanks for any suggestions!

**Comment by josegonzalez** [comment]
You have at least one application - although possibly many - that are failing to deploy with a single ip address associated with it. As such, I would like you to provide the `dokku report APP_NAME` output for that application, as well as the logs of a deploy after running `dokku trace on`. Without that, I do not have any visibility into what options Dokku is seeing.

**Comment by babakness** [comment]
Hey, Jose, thanks for following up. 

I believe my question is being mis-understood. They all deploy perfectly fine, its just that they are "greedy" and BIND to ANY ip they can get there hands on running on the host.

See, if I point the A record of a domain to one IP address going into the host, it works. If I use a different IP for the A record, it also works. But I only want ONE ip to work not ALL.

The output of such an app is given in this issue:

https://github.com/dokku/dokku/issues/3191

I've been sweating this issue for hours. Please provide any insight you may have. I was able to confirm with someone else that `daemon.json` does indeed confined docker containers to some ip given that `daemon.json` provides the ip as follows:

```
{
  "ip": "172.30.0.10"
}
```

In this case, the services only listen to that specific IP, this isn't the case for the `dokku` launched containers (which have info that looks like the issue aforementioned)

**Comment by josegonzalez** [comment]
Can you please provide the output of a deploy after running `dokku trace on`. I understand what you are trying to do - as well as what is going on - but I _need_ to be sure that certain things are happening in the wrong/right order.

**Comment by babakness** [comment]
OK

```
+ export DOKKU_HOST_ROOT=/home/dokku
+ DOKKU_HOST_ROOT=/home/dokku
+ export DOKKU_DISTRO
++ . /etc/os-release
++ echo ubuntu
+ DOKKU_DISTRO=ubuntu
+ export DOKKU_IMAGE=gliderlabs/herokuish
+ DOKKU_IMAGE=gliderlabs/herokuish
+ export DOKKU_LIB_ROOT=/var/lib/dokku
+ DOKKU_LIB_ROOT=/var/lib/dokku
+ export PLUGIN_PATH=/var/lib/dokku/plugins
+ PLUGIN_PATH=/var/lib/dokku/plugins
+ export PLUGIN_AVAILABLE_PATH=/var/lib/dokku/plugins/available
+ PLUGIN_AVAILABLE_PATH=/var/lib/dokku/plugins/available
+ export PLUGIN_ENABLED_PATH=/var/lib/dokku/plugins/enabled
+ PLUGIN_ENABLED_PATH=/var/lib/dokku/plugins/enabled
+ export PLUGIN_CORE_PATH=/var/lib/dokku/core-plugins
+ PLUGIN_CORE_PATH=/var/lib/dokku/core-plugins
+ export PLUGIN_CORE_AVAILABLE_PATH=/var/lib/dokku/core-plugins/available
+ PLUGIN_CORE_AVAILABLE_PATH=/var/lib/dokku/core-plugins/available
+ export PLUGIN_CORE_ENABLED_PATH=/var/lib/dokku/core-plugins/enabled
+ PLUGIN_CORE_ENABLED_PATH=/var/lib/dokku/core-plugins/enabled
+ export DOKKU_API_VERSION=1
+ DOKKU_API_VERSION=1
+ export DOKKU_NOT_IMPLEMENTED_EXIT=10
+ DOKKU_NOT_IMPLEMENTED_EXIT=10
+ export DOKKU_VALID_EXIT=0
+ DOKKU_VALID_EXIT=0
+ export DOKKU_LOGS_DIR=/var/log/dokku
+ DOKKU_LOGS_DIR=/var/log/dokku
+ export DOKKU_EVENTS_LOGFILE=/var/log/dokku/events.log
+ DOKKU_EVENTS_LOGFILE=/var/log/dokku/events.log
+ export DOKKU_CONTAINER_LABEL=dokku
+ DOKKU_CONTAINER_LABEL=dokku
+ export DOKKU_GLOBAL_RUN_ARGS=--label=dokku
+ DOKKU_GLOBAL_RUN_ARGS=--label=dokku
+ source /var/lib/dokku/core-plugins/available/common/functions
++ set -eo pipefail
++ [[ -n 1 ]]
++ set -x
+ parse_args git-receive-pack ''\''node-sample'\'''
+ declare 'desc=top-level cli arg parser'
+ local next_index=1
+ local skip=false
+ args=("$@")
+ local args
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ next_index=2
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ next_index=3
+ return 0
+ args=("$@")
+ skip_arg=false
+ [[ git-receive-pack =~ ^--.* ]]
+ has_tty
+ declare 'desc=return 0 if we have a tty'
++ /usr/bin/tty
++ true
+ [[ not a tty == \n\o\t\ \a\ \t\t\y ]]
+ return 1
+ DOKKU_QUIET_OUTPUT=1
++ id -un
+ [[ dokku != \d\o\k\k\u ]]
+ [[ git-receive-pack =~ ^plugin:.* ]]
+ [[ git-receive-pack == \s\s\h\-\k\e\y\s\:\a\d\d ]]
+ [[ -n git-receive-pack 'node-sample' ]]
+ export -n SSH_ORIGINAL_COMMAND
+ [[ git-receive-pack =~ config-* ]]
+ [[ git-receive-pack =~ docker-options* ]]
+ set -f
+ /usr/bin/dokku git-receive-pack ''\''node-sample'\'''
+ export DOKKU_HOST_ROOT=/home/dokku
+ DOKKU_HOST_ROOT=/home/dokku
+ export DOKKU_DISTRO
++ . /etc/os-release
++ echo ubuntu
+ DOKKU_DISTRO=ubuntu
+ export DOKKU_IMAGE=gliderlabs/herokuish
+ DOKKU_IMAGE=gliderlabs/herokuish
+ export DOKKU_LIB_ROOT=/var/lib/dokku
+ DOKKU_LIB_ROOT=/var/lib/dokku
+ export PLUGIN_PATH=/var/lib/dokku/plugins
+ PLUGIN_PATH=/var/lib/dokku/plugins
+ export PLUGIN_AVAILABLE_PATH=/var/lib/dokku/plugins/available
+ PLUGIN_AVAILABLE_PATH=/var/lib/dokku/plugins/available
+ export PLUGIN_ENABLED_PATH=/var/lib/dokku/plugins/enabled
+ PLUGIN_ENABLED_PATH=/var/lib/dokku/plugins/enabled
+ export PLUGIN_CORE_PATH=/var/lib/dokku/core-plugins
+ PLUGIN_CORE_PATH=/var/lib/dokku/core-plugins
+ export PLUGIN_CORE_AVAILABLE_PATH=/var/lib/dokku/core-plugins/available
+ PLUGIN_CORE_AVAILABLE_PATH=/var/lib/dokku/core-plugins/available
+ export PLUGIN_CORE_ENABLED_PATH=/var/lib/dokku/core-plugins/enabled
+ PLUGIN_CORE_ENABLED_PATH=/var/lib/dokku/core-plugins/enabled
+ export DOKKU_API_VERSION=1
+ DOKKU_API_VERSION=1
+ export DOKKU_NOT_IMPLEMENTED_EXIT=10
+ DOKKU_NOT_IMPLEMENTED_EXIT=10
+ export DOKKU_VALID_EXIT=0
+ DOKKU_VALID_EXIT=0
+ export DOKKU_LOGS_DIR=/var/log/dokku
+ DOKKU_LOGS_DIR=/var/log/dokku
+ export DOKKU_EVENTS_LOGFILE=/var/log/dokku/events.log
+ DOKKU_EVENTS_LOGFILE=/var/log/dokku/events.log
+ export DOKKU_CONTAINER_LABEL=dokku
+ DOKKU_CONTAINER_LABEL=dokku
+ export DOKKU_GLOBAL_RUN_ARGS=--label=dokku
+ DOKKU_GLOBAL_RUN_ARGS=--label=dokku
+ source /var/lib/dokku/core-plugins/available/common/functions
++ set -eo pipefail
++ [[ -n 1 ]]
++ set -x
+ parse_args git-receive-pack ''\''node-sample'\'''
+ declare 'desc=top-level cli arg parser'
+ local next_index=1
+ local skip=false
+ args=("$@")
+ local args
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ next_index=2
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ next_index=3
+ return 0
+ args=("$@")
+ skip_arg=false
+ [[ git-receive-pack =~ ^--.* ]]
+ has_tty
+ declare 'desc=return 0 if we have a tty'
++ /usr/bin/tty
++ true
+ [[ not a tty == \n\o\t\ \a\ \t\t\y ]]
+ return 1
+ DOKKU_QUIET_OUTPUT=1
++ id -un
+ [[ dokku != \d\o\k\k\u ]]
+ [[ git-receive-pack =~ ^plugin:.* ]]
+ [[ git-receive-pack == \s\s\h\-\k\e\y\s\:\a\d\d ]]
+ [[ -n '' ]]
+ dokku_auth git-receive-pack ''\''node-sample'\'''
+ declare 'desc=calls user-auth plugin trigger'
+ export SSH_USER=dokku
+ SSH_USER=dokku
+ export SSH_NAME=admin1
+ SSH_NAME=admin1
+ plugn trigger user-auth dokku admin1 git-receive-pack ''\''node-sample'\'''
+ source /var/lib/dokku/core-plugins/available/common/functions
++ set -eo pipefail
++ [[ -n 1 ]]
++ set -x
+ [[ ! -n '' ]]
+ source /var/lib/dokku/core-plugins/available/common/functions
++ set -eo pipefail
++ [[ -n 1 ]]
++ set -x
+ [[ git-receive-pack =~ (ssh-keys:add|ssh-keys:remove) ]]
+ return 0
+ case "$1" in
+ execute_dokku_cmd git-receive-pack ''\''node-sample'\'''
+ declare 'desc=executes dokku sub-commands'
+ local PLUGIN_NAME=git-receive-pack
+ local PLUGIN_CMD=git-receive-pack
+ local implemented=0
+ local script
+ argv=("$@")
+ local argv
+ case "$PLUGIN_NAME" in
++ readlink -f /var/lib/dokku/plugins/enabled/git-receive-pack
+ [[ /var/lib/dokku/plugins/enabled/git-receive-pack == *core-plugins* ]]
+ [[ -x /var/lib/dokku/plugins/enabled/git-receive-pack/subcommands/default ]]
+ [[ -x /var/lib/dokku/plugins/enabled/git-receive-pack/subcommands/git-receive-pack ]]
+ [[ -x /var/lib/dokku/plugins/enabled/git-receive-pack/subcommands/git-receive-pack ]]
+ [[ 0 -eq 0 ]]
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/00_dokku-standard/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/20_events/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/apps/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/certs/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/checks/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/config/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/docker-options/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/domains/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/enter/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/git/commands git-receive-pack ''\''node-sample'\'''
+ source /var/lib/dokku/plugins/available/git/functions
++ source /var/lib/dokku/core-plugins/available/common/functions
+++ set -eo pipefail
+++ [[ -n 1 ]]
+++ set -x
++ source /var/lib/dokku/core-plugins/available/common/property-functions
+++ set -eo pipefail
+++ [[ -n 1 ]]
+++ set -x
++ source /var/lib/dokku/plugins/available/apps/functions
+++ set -eo pipefail
+++ [[ -n 1 ]]
+++ set -x
+++ source /var/lib/dokku/core-plugins/available/common/functions
++++ set -eo pipefail
++++ [[ -n 1 ]]
++++ set -x
+++ source /var/lib/dokku/plugins/available/config/functions
++++ set -eo pipefail
++++ [[ -n 1 ]]
++++ set -x
++++ source /var/lib/dokku/core-plugins/available/common/functions
+++++ set -eo pipefail
+++++ [[ -n 1 ]]
+++++ set -x
++ source /var/lib/dokku/plugins/available/config/functions
+++ set -eo pipefail
+++ [[ -n 1 ]]
+++ set -x
+++ source /var/lib/dokku/core-plugins/available/common/functions
++++ set -eo pipefail
++++ [[ -n 1 ]]
++++ set -x
++ set -eo pipefail
++ [[ -n 1 ]]
++ set -x
+ source /var/lib/dokku/plugins/available/git/internal-functions
++ source /var/lib/dokku/core-plugins/available/common/functions
+++ set -eo pipefail
+++ [[ -n 1 ]]
+++ set -x
++ source /var/lib/dokku/core-plugins/available/common/property-functions
+++ set -eo pipefail
+++ [[ -n 1 ]]
+++ set -x
++ set -eo pipefail
++ [[ -n 1 ]]
++ set -x
+ case "$1" in
+ git_glob_cmd git-receive-pack ''\''node-sample'\'''
+ declare 'desc=catch-all for any other git-* commands'
+ local 'cmd=git-*'
++ sed 's/^\///g'
++ sed 's/\\'\''/'\''/g'
++ perl -pe 's/(?<!\\)'\''//g'
++ echo ''\''node-sample'\'''
+ local APP=node-sample
+ local APP_PATH=/home/dokku/node-sample
+ is_valid_app_name node-sample
+ declare 'desc=verify app name format'
+ local APP=node-sample
+ [[ ! -n node-sample ]]
+ [[ node-sample =~ ^[a-z].* ]]
+ [[ ! node-sample =~ [A-Z] ]]
+ [[ ! node-sample =~ [:] ]]
+ return 0
+ [[ git-receive-pack == \g\i\t\-\r\e\c\e\i\v\e\-\p\a\c\k ]]
+ [[ ! -d /home/dokku/node-sample/refs ]]
+ [[ git-receive-pack == \g\i\t\-\r\e\c\e\i\v\e\-\p\a\c\k ]]
+ local 'args=git-receive-pack '\''/home/dokku/node-sample'\'''
+ git-shell -c 'git-receive-pack '\''/home/dokku/node-sample'\'''
+ exit_code=0
+ set -e
+ [[ 0 -eq 10 ]]
+ implemented=1
+ [[ 0 -ne 0 ]]
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/letsencrypt/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/logs/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/network/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/nginx-vhosts/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/plugin/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/postgres/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/proxy/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/ps/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/repo/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/shell/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/ssh-keys/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/storage/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/tags/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ for script in $PLUGIN_ENABLED_PATH/*/commands
+ set +e
+ /var/lib/dokku/plugins/enabled/tar/commands git-receive-pack ''\''node-sample'\'''
+ exit_code=10
+ set -e
+ [[ 10 -eq 10 ]]
+ continue
+ [[ 1 -eq 0 ]]
+ set +f
+ exit 0
Everything up-to-date
```

**Comment by josegonzalez** [comment]
If you turn off nginx, are you able to route to those containers? I'm going to guess that it is actually nginx that is listening to all interfaces and not docker containers. By default, we run docker containers on internal ip addresses - not bound to `0.0.0.0` - and the command being run here confirms that we didn't use the `-P` flag:

```
docker run --label=dokku \
			-d \
			-e PORT=5000 \
			--env=USER=herokuishuser \
			--restart=on-failure:10 \
			-e DYNO=web.1 \
			-e TRACE=true \
			dokku/node-sample:latest /start web
```

As far as binding nginx to a specific ip address, we don't support configuring the default ip address it binds to as a simple `nginx:set` option - it's all or nothing - but you should be able to configure it by providing your application with a custom `nginx.conf.sigil`. You can see documentation for that [here](http://dokku.viewdocs.io/dokku/configuration/nginx/#customizing-the-nginx-configuration).

Let me know if the above sounds right. Given that it's a herokuish app, you aren't publishing ports, and nginx is being used, I'm pretty confident it is, but I could be wrong.

**Comment by babakness** [comment]
Interesting. I do have `letsencrypt`, not sure if that might be doing something to wrap access to the containers... I'll try out your suggestion

**Comment by josegonzalez** [comment]
The letsencrypt plugin only inserts an extra stanza for validating ssl to the nginx conf - it gets removed after reuse - and doesn't otherwise change what nginx is listening to. Nginx is our default routing layer, and is enabled by default - it's enabled for this app, which can be verified from the output of `dokku report APP_NAME`.

**Comment by babakness** [comment]
Stopping nginx does indeed stop all traffic. Can I manually have nginx listen to a specific IP? I noticed that `sites-available` has a configuration to listen for all IPs but it isn't in `sites-enabled`. Searching for the `listen` directive I find

```
nginx.conf:#		listen     localhost:110;
nginx.conf:#		listen     localhost:143;
sites-available/default:	listen 80 default_server;
sites-available/default:	listen [::]:80 default_server;
sites-available/default:	# listen 443 ssl default_server;
sites-available/default:	# listen [::]:443 ssl default_server;
sites-available/default:#	listen 80;
sites-available/default:#	listen [::]:80;
```

Where might I be able to restrict it?

**Comment by josegonzalez** [comment]
Those default sites have nothing to do with how Dokku routes using nginx.

As I mentioned before, there is no way for you to globally set the nginx configuration for all apps to listen to a single ip address. It's not a common enough thing for us to have anything for it at this time.

Some alternatives to do what you want:

- You can provide a custom [nginx.conf.sigil](http://dokku.viewdocs.io/dokku/configuration/nginx/#customizing-the-nginx-configuration) with each application.
- You can inject a custom `nginx.conf.sigil` file using the `post-extract` plugin trigger in a custom plugin. Docs on creating plugins are [here](http://dokku.viewdocs.io/dokku/development/plugin-creation/)
- You can provide a custom nginx plugin based on our own that customizes the ip address in it's default template. This is probably the most time-consuming option because each trigger checks to see if it should be enabled, and you still need to set the proxy method to your custom routing plugin on a per-app basis.

**Comment by babakness** [comment]
I see, so I would check in a file named nginx.conf.sigil at the root of my repo which is then parsed to create the nginx.conf for the app. I'll try that...

**Comment by babakness** [comment]
Is it possible set a global `dokku` variable that contains the desired IP that is included into the `sigil` template?

**Comment by josegonzalez** [comment]
No, thats not currently possible. At this time, the nginx plugin doesn't have a way to set custom metadata for the template.

**Comment by babakness** [comment]
Got it. Is it just a limitation of that nginx plugin? Is it possible to pass down a variable to plugins globally? For example, to issue a re-build command to all containers which would then use the environment variables to rebuild from the original `sigil` template?

**Comment by babakness** [comment]
An interesting idea is to create a hostname in `/etc/hosts` which can be set to whatever IP is desired without having to go through passing variables or re-compiling.

**Comment by babakness** [comment]
For example
```
server {
  listen special-dokku-host:443
}
```

**Comment by babakness** [comment]
Anyway, thanks for the help! I learned more about `dokku` today, great project, glad I found it!

**Comment by babakness** [comment]
Correction @josegonzalez 

`post-deploy` will work here. `post-extract` will not.

**Comment by babakness** [comment]
Ah, I understand what you meant now. My misunderstanding of how things work and what you stated. I did go for a modified version of the second step you suggested above. 

I thought to myself that the default `nginx.conf.sigil` that gets shipped may get modified or enhanced in future updates; there is also the possibility someone makes mods wanting to add specific customizations into nginx by using the first technique you mentioned where one checks in a `nginx.conf.sigil` into the base directory. So I figured the solution would better tolerate enhancements or customizations to substitute the `listen` directive only and after it had been processed. 

In short I'm not injecting a custom `nginx.conf.sigil` that would override the functionality of someone checking one in themselves; instead I'm only parsing the listen directive in the final output. One for IPv4 and one for IPv6; on both ports 80 and 443.

**Comment by babakness** [comment]
@josegonzalez 

New problem:  Rebooting the server reverts all changes.

I'm using all the following trigger for good measure, but they fail the reboot test
1. `post-extract`
2. `post-deploy`
3. `pre-start`

Basically when you reboot, the original `nginx.conf.sigil` is parsed and injected into the app.

**Comment by josegonzalez** [comment]
I don't know what the code it is you wrote so I cannot tell you whether or not it will work. This is also not the best forum for this sort of support, and I recommend you join us on slack with further questions - we can update this thread once we figure out what is going on.

**Comment by babakness** [comment]
Sounds good, I've posted on Slack.
